### PR TITLE
[FrameworkBundle] add --deprecations on debug:container command

### LINKED ISSUE #30089: Display deprecated triggered during cache warmup
Hi!

**Description**  
Currently it does not look possible to get deprecated calls generated during cache warmup.
These deprecations look to be stored in a file like `var/cache/dev/srcApp_KernelDevDebugContainerDeprecations.log` and they are displayed in the web profiler.
But when you have a CLI only app, one could like to see these deprecations directly in CLI (better DX IMHO).

I think that it would be interesting to display these deprecations during `cache:warmup` when a very verbose mode is activated like `-vvv`.

Also, what about a new command that would display these cached deprecations? Something like `debug:deprecations` or `debug:cached-deprecations`.

Cheers!

### PR DESCRIPTION
**Pull request by noemi-salaun**
| Q             | A
| ------------- | ---
| Branch?       | master (5.1)
| Bug fix?      | no
| New feature?  | yes 
| Deprecations? | no 
| Tickets       | #30089  
| License       | MIT
| Doc PR        | todo

_I apologize in advance, I am not fluent in English_

Continuity of @Simperfit work from his PR https://github.com/symfony/symfony/pull/32584

I added support for XML, JSON and markdown formats as well as unit tests for all formats.

## XML format

```xml
<?xml version="1.0" encoding="UTF-8"?>
<deprecations remainingCount="5">
  <deprecation count="3">
    <message>Some deprecation message.</message>
    <file>/path/to/some/file.php</file>
    <line>39</line>
  </deprecation>
  <deprecation count="2">
    <message>An other deprecation message.</message>
    <file>/path/to/an/other/file.php</file>
    <line>25</line>
  </deprecation>
</deprecations>
```

## JSON format

```json
{
  "remainingCount": 5,
  "deprecations": [
    {
      "message": "Some deprecation message.",
      "file": "\/path\/to\/some\/file.php",
      "line": 39,
      "count": 3
    },
    {
      "message": "An other deprecation message.",
      "file": "\/path\/to\/an\/other\/file.php",
      "line": 25,
      "count": 2
    }
  ]
}
```

## Remaining deprecations (5)

- 3x: "Some deprecation message." in /path/to/some/file.php:39
- 2x: "An other deprecation message." in /path/to/an/other/file.php:25

I added a new commit on top of @Simperfit 's one, but I don't know how you would like to have the git history writed.

### CODE REVIEW

**Comment by chalasr** [comment]
Thanks for the taking over!
Features are for the master branch :) Rebase needed

**Comment by noemi-salaun** [comment]
Oh ok, I was targeting the same version as the previous PR but I guess 5.0 was release since ^^

**Comment by noemi-salaun** [comment]
Oh I rebased before switching the targeted branch for this PR, so the CI failed

**Comment by maxhelias** [comment]
The history of commits is weird, can you take a look for remove the commit alreay merge

**Comment by noemi-salaun** [comment]
I have rebased Simperfit branch on top of 4.4 branch, then rebase again on top of master. What do you suggest @maxhelias ? I restart from master and cherry pick the commit from Simperfit and mine ?

File: src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md [hunk 1]
```diff
@@ -13,6 +13,7 @@ CHANGELOG
  * The `TemplateController` now accepts context argument
  * Deprecated *not* setting the "framework.router.utf8" configuration option as it will default to `true` in Symfony 6.0
  * Added tag `routing.expression_language_function` to define functions available in route conditions
+ * Added `debug:container --deprecations` command to see compile-time deprecations.
```
Comment (noniagriconomie) [inline src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md]: s/command/option

File: src/Symfony/Bundle/FrameworkBundle/Command/ContainerDebugCommand.php [hunk 2]
```diff
@@ -62,13 +62,18 @@ protected function configure()
                 new InputOption('env-vars', null, InputOption::VALUE_NONE, 'Displays environment variables used in the container'),
                 new InputOption('format', null, InputOption::VALUE_REQUIRED, 'The output format (txt, xml, json, or md)', 'txt'),
                 new InputOption('raw', null, InputOption::VALUE_NONE, 'To output raw description'),
+                new InputOption('deprecations', null, InputOption::VALUE_NONE, 'To output the deprecations generated when compiling and warming the cache'),
```
Comment (noniagriconomie) [inline src/Symfony/Bundle/FrameworkBundle/Command/ContainerDebugCommand.php]: The container no?

File: src/Symfony/Bundle/FrameworkBundle/Command/ContainerDebugCommand.php [hunk 3]
```diff
@@ -62,13 +62,18 @@ protected function configure()
...
             ])
             ->setDescription('Displays current services for an application')
             ->setHelp(<<<'EOF'
 The <info>%command.name%</info> command displays all configured <comment>public</comment> services:
 
   <info>php %command.full_name%</info>
 
+To see deprecations generated during container compilation and cache warmup, use the <info>--deprecations</info> flag:
```
Comment (noniagriconomie) [inline src/Symfony/Bundle/FrameworkBundle/Command/ContainerDebugCommand.php]: Maybe option instead of flag would be better description

**Review by noniagriconomie — COMMENTED** [review]
Wanted to give it a try also, you beat me at this
Thank you continuing this, i think it will be very usefull

File: src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md (hunk 1 above)
Comment (noemi-salaun) [reply src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md]: I didn't write that part but I can make the change you are suggesting :)

File: src/Symfony/Bundle/FrameworkBundle/Command/ContainerDebugCommand.php (hunk 2 above)
Comment (noemi-salaun) [reply src/Symfony/Bundle/FrameworkBundle/Command/ContainerDebugCommand.php]: I didn't write that part but I can make the change you are suggesting :)

```suggestion
                new InputOption('deprecations', null, InputOption::VALUE_NONE, 'To output the deprecations generated when compiling the container and warming up the cache'),
```

File: src/Symfony/Bundle/FrameworkBundle/Command/ContainerDebugCommand.php (hunk 3 above)
Comment (noemi-salaun) [reply src/Symfony/Bundle/FrameworkBundle/Command/ContainerDebugCommand.php]: The other descriptions use the 2 words without distinction. `--env-var` is sometime called a flag and other time an option :/

```
To see available types that can be used for autowiring, use the <info>--types</info> flag:

  <info>php %command.full_name% --types</info>

To see environment variables used by the container, use the <info>--env-vars</info> flag:

  <info>php %command.full_name% --env-vars</info>

Display a specific environment variable by specifying its name with the <info>--env-var</info> option:

  <info>php %command.full_name% --env-var=APP_ENV</info>
```

I'll make the change ;)

Comment (noniagriconomie) [reply src/Symfony/Bundle/FrameworkBundle/Command/ContainerDebugCommand.php]: oki understood, lets wait other reviews 👍

File: src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/Descriptor.php [hunk 4]
```diff
@@ -132,6 +135,11 @@ protected function write(string $content, bool $decorated = false)
      */
     abstract protected function describeContainerServices(ContainerBuilder $builder, array $options = []);
 
+    /**
+     * Describes container deprecations.
+     */
+    abstract protected function describeContainerDeprecations(ContainerBuilder $builder, array $options = []);
```
Comment (ogizanagi) [inline src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/Descriptor.php]: Missing `: void` return type?

**Review by ogizanagi — APPROVED** [review]
Thank you for taking over this

**Review by maxhelias — APPROVED** [review]
Great! it's look good to me

File: src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/Descriptor.php (hunk 4 above)
Comment (noemi-salaun) [reply src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/Descriptor.php]: The others `describeContainerXXX` already defined in this file does not declare a `void` return type. I would prefer to remain consistent with the whole file.

Comment (ogizanagi) [reply src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/Descriptor.php]: But you declared it in implementations. So it should be consistent with the interface IMHO.

Comment (noemi-salaun) [reply src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/Descriptor.php]: Oh you are right, I did not see that. The method declarations was already added in Simperfit commit. Would you prefer `void` return type on the interface, or I remove it from the implementations ?

File: src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/Descriptor.php [hunk 5]
```diff
@@ -132,6 +135,11 @@ protected function write(string $content, bool $decorated = false)
      */
     abstract protected function describeContainerServices(ContainerBuilder $builder, array $options = []);
 
+    /**
+     * Describes container deprecations.
+     */
```
Comment (chalasr) [inline src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/Descriptor.php]: Can be removed.

File: src/Symfony/Bundle/FrameworkBundle/Command/ContainerDebugCommand.php [hunk 6]
```diff
@@ -180,7 +187,7 @@ protected function execute(InputInterface $input, OutputInterface $output): int
                 $errorIo->comment('To search for a specific tag, re-run this command with a search term. (e.g. <comment>debug:container --tag=form.type</comment>)');
             } elseif ($input->getOption('parameters')) {
                 $errorIo->comment('To search for a specific parameter, re-run this command with a search term. (e.g. <comment>debug:container --parameter=kernel.debug</comment>)');
-            } else {
+            } elseif (!$input->getOption('deprecations')) {
```
Comment (chalasr) [inline src/Symfony/Bundle/FrameworkBundle/Command/ContainerDebugCommand.php]: The `validateInput` method below should be updated to take the new option into account.

**Review by chalasr — CHANGES_REQUESTED** [review]

File: src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/Descriptor.php (hunk 5 above)
Comment (noemi-salaun) [reply src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/Descriptor.php]: I can remove it but all the other `describeContainerXXX` have a short description that mirror the method name.

```php
    /**
     * Describes a service definition.
     */
    abstract protected function describeContainerDefinition(Definition $definition, array $options = []);

    /**
     * Describes a service alias.
     */
    abstract protected function describeContainerAlias(Alias $alias, array $options = [], ContainerBuilder $builder = null);

    /**
     * Describes a container parameter.
     */
    abstract protected function describeContainerParameter($parameter, array $options = []);
```

@chalasr Should I remove all the redundant description or only the description for the new `describeContainerDeprecations()` method ?

File: src/Symfony/Bundle/FrameworkBundle/Command/ContainerDebugCommand.php (hunk 6 above)
Comment (noemi-salaun) [reply src/Symfony/Bundle/FrameworkBundle/Command/ContainerDebugCommand.php]: Oh right ! Fixed in the new version :)

File: src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/Descriptor.php (hunk 5 above)
Comment (chalasr) [reply src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/Descriptor.php]: Only the new one please. The remaining should be removed on the lowest maintained branch where they exist.

File: src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/Descriptor.php (hunk 4 above)
Comment (ogizanagi) [reply src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/Descriptor.php]: Let's add it. It's accurate and does not restrict anything anyway as this class is internal.

Comment (noemi-salaun) [reply src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/Descriptor.php]: Ok done :)

File: src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/Descriptor.php (hunk 5 above)
Comment (noemi-salaun) [reply src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/Descriptor.php]: Ok done :)

**Review by SelviA — APPROVED** [review]

**Comment by noemi-salaun** [comment]
I think @dunglas, @jderusse, @lyrixx, @sroze and @xabbuh were automatically added as reviewer when I strangely rebase between branch 4.4 and master. The PR contained their commit by mistake. But I cannot remove them from the reviewers... I'm sorry for the flood :disappointed:

**Review by chalasr — APPROVED** [review]

File: src/Symfony/Bundle/FrameworkBundle/Tests/Functional/ContainerDebugCommandTest.php [hunk 7]
```diff
@@ -136,6 +136,73 @@ public function testDescribeEnvVar()
...
+        @unlink(static::$container->getParameter('debug.container.dump'));
+
+        $tester = new ApplicationTester($application);
+        $tester->run(['command' => 'debug:container', '--deprecations' => true]);
+
+        $this->assertSame(0, $tester->getStatusCode());
+        $this->assertStringContainsString('[OK] There are no deprecations in the logs!', $tester->getDisplay());
+    }
+
+    public function testGetDeprecationNoFile(): void
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Functional/ContainerDebugCommandTest.php]: `void` should be removed

**Review by nicolas-grekas — APPROVED** [review]
Thanks @noemi-salaun for taking over!

**Comment by nicolas-grekas** [comment]
Thank you @noemi-salaun.

**Comment by nicolas-grekas** [comment]
And thank you @Simperfit of course also!

**Pull request merged**